### ISSUE TITLE
official 1.15.0 gpu image on dockerhub 10x slower than 1.14.0

### ISSUE THREAD

**Issue by ilaripih** [tensorflow/serving#1469]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04
- **TensorFlow Serving installed from (source or binary)**: binary (official Dockerhub image)
- **TensorFlow Serving version**: 1.15.0-gpu
- **GPU**: NVIDIA GeForce GTX 1080 (8GB)

### Describe the problem
I use the Retinanet object detection model with Resnet 50 backbone as defined here: https://github.com/tensorflow/models/blob/master/research/object_detection/samples/configs/ssd_resnet50_v1_fpn_shared_box_predictor_640x640_coco14_sync.config
With the official `tensorflow/serving:1.14.0-gpu` dockerhub image I could run the model at 20 - 30 FPS depending on batch size. When I upgraded to 1.15.0-rc2 the speed dropped to about 2.7 FPS but the GPU utilization stayed at 100%. Changing the batch size had very little effect on the speed.

When I upgraded again to the nightly build (`tensorflow/serving:nightly-gpu`) last week (before the final 1.15.0 release) the performance issue was resolved. Then the final 1.15.0 release happened, I upgraded to it, and again I got only 2.7 FPS (with 100% GPU utilization). Today, I once again upgraded to the nightly-gpu version and get a steady 30 FPS again.

The same performance issue happened on another machine with a GTX 1060 (6GB) card. I'm wondering if the tagged releases are built differently for Dockerhub than the nighly builds because currently the nightly builds deliver a good performance and the tagged releases are 10x slower.

### Exact Steps to Reproduce
Pull the `tensorflow/serving:1.15.0-gpu` Dockerhub image, run the Retinanet model with Resnet 50 backbone on it, notice terrible performance compared to version 1.14.0 or any nightly build.

**Comment by gowthamkpr** [comment]
@ilaripih Was it just happening with Retinanet model or is it happening with all the models? Thanks!

**Comment by ilaripih** [comment]
It just happens with the Retinanet model (I also tried Deeplab and InceptionV4). But it gets weirder: I set up a test environment here: https://github.com/ilaripih/tensorflow-serving-performance-test
And I noticed that I have no issues with the official `ssd_resnet50_v1_fpn_shared_box_predictor_640x640_coco14_sync_2018_07_03.tar.gz` checkpoint (more specifically the SavedModel inside it) from the tensorflow/models repo; the performance is even a bit better with `serving:1.15.0-gpu` than with `serving:1.14.0-gpu`. The performance degradation happens with our custom Retinanet model that only has about 15 labels and was trained with Tensorflow 1.14.0.

The Serving signature of the model looks like this:
```
MetaGraphDef with tag-set: 'serve' contains the following SignatureDefs:

signature_def['labels']:
  The given SavedModel SignatureDef contains the following input(s):
  The given SavedModel SignatureDef contains the following output(s):
    outputs['labels'] tensor_info:
        dtype: DT_STRING
        shape: (1)
        name: labels:0
  Method name is: tensorflow/serving/predict

signature_def['serving_default']:
  The given SavedModel SignatureDef contains the following input(s):
    inputs['inputs'] tensor_info:
        dtype: DT_UINT8
        shape: (-1, 640, 640, 3)
        name: image_tensor:0
  The given SavedModel SignatureDef contains the following output(s):
    outputs['detection_boxes'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 100, 4)
        name: detection_boxes:0
    outputs['detection_classes'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 100)
        name: detection_classes:0
    outputs['detection_multiclass_scores'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 100, 13)
        name: detection_multiclass_scores:0
    outputs['detection_scores'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1, 100)
        name: detection_scores:0
    outputs['num_detections'] tensor_info:
        dtype: DT_FLOAT
        shape: (-1)
        name: num_detections:0
  Method name is: tensorflow/serving/predict
```

What's different from the official model is that we have a `labels` SignatureDef (which is not used in this performance test) in addition to `serving_default`, and the input height and width are fixed whereas they're variable in the official model. Our Deeplab and InceptionV4 models also have the same `labels` SignatureDef and fixed-size input (except for batch-size) but they're running just fine.

**Comment by gowthamkpr** [comment]
Is there a chance thet you can provide your custom Retinanet model as with the above information its hard for us to identify the root cause of the issue. Thanks!

**Comment by ilaripih** [comment]
Yes, I'll try to get a minimum reproducible model later today.

**Comment by dave-andersen** [comment]
Btw - seeing exactly this problem using a nearly off-the-shelf SSD FPN mobilenets model using the pip install of TF 1.15 (executing inside an nvidia-docker image but I'm not certain that's relevant -- with tf 1.14 in the same docker image I get normal speed).  About a 2.7x slowdown -- my test run goes from 108 seconds to 280 seconds.

This is a fine-tuned SSD mobilenets FPN coco config, with the only changes being:
 - Fixed size input of 1024x1024
 - Single class detection

I've attached my pipeline.config in case it's helpful for reproducing this bug.
[pipeline.config.txt](https://github.com/tensorflow/serving/files/3811221/pipeline.config.txt)

**Comment by dave-andersen** [comment]
(In thinking about this, I suspect a performance regression in 1.15, not a serving-specific slowdown.  I'm just running natively in TF with my own inference pipeline, doing the obvious feed_dict to get data in:

```
            out = sess.run(
                [
                    sess.graph.get_tensor_by_name(node) for node in [
                        'num_detections:0', 'detection_scores:0',
                        'detection_boxes:0', 'detection_classes:0'
                    ]
                ],
                feed_dict={'image_tensor:0': img_batch})
```

Oh - worth noting the other change to SSD FPN, that model draws from layers 2-5 instead of 3-6 and so is a fair bit larger than one might otherwise expect.

**Comment by gowthamkpr** [comment]
@dave-andersen This definitely looks like a regression issue in TF1.15. Can you please post this issue in [tensorflow/tensorflow](https://github.com/tensorflow/tensorflow) and [tensorflow/models](https://github.com/tensorflow/models) repo. Thanks!

@ilaripih Can you please provide a minimal example for us to reproduce. Thanks!

**Comment by ilaripih** [comment]
I finally had the time to find out how to easily reproduce the slow (on Serving 1.15.0-gpu) SavedModel:

Either download the SavedModel from here: https://drive.google.com/file/d/1VIZ3_YFOCVI3-txYwAN_ApxtgDCm3Bgm/view?usp=sharing

...or if you want to create it yourself:

1. Install the CPU version of Tensorflow 1.14.0. It's possible that this works with any Tensorflow version but I haven't tested.
2. Clone the master branch of the tensorflow/models repo
3. Download and extract the official Retinanet model's checkpoint files: http://download.tensorflow.org/models/object_detection/ssd_resnet50_v1_fpn_shared_box_predictor_640x640_coco14_sync_2018_07_03.tar.gz
4. Re-export the SavedModel by running the `export_inference_graph.py` script with `--input_shape=-1,640,640,3`.

So all one has to do is export the model with the original pipeline config.

Now that the 2.0.0-gpu release is out I tested it too, and the performance is close to 1.14.0-gpu. So the performance degradation seems to only affect the 1.15.0-gpu version.

**Comment by ilaripih** [comment]
This is definitely a Tensorflow issue because the same performance degradation can be reproduced by running the SavedModel directly with libtensorflow-gpu 1.15.0.

**Comment by gowthamkpr** [comment]
@ilaripih I am closing this issue as it looks like the issue is being caused by tensorflow-gpu 1.1. itself andis being tracked [here](https://github.com/tensorflow/tensorflow/issues/34174). Thanks!

**Comment by asgg4321** [comment]
I train custom model in 1.14.0, and then, I used tf.saved_model.simple_save() to export tf.keras.models.Model as savedmodel. It's slowly for load model in tensorflow/serving 2.0 than tensorflow/serving 1.14.0.  I don't know if this is the same as your problem. Did you solve the problem?
